crates/catalog/inmemory/src/catalog.rs

```diff
@@ -0,0 +1,1496 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! This module contains in-memory catalog implementation.
+
+use futures::lock::Mutex;
+use iceberg::io::FileIO;
+use iceberg::spec::{TableMetadata, TableMetadataBuilder};
+use itertools::Itertools;
+use std::collections::HashMap;
+use uuid::Uuid;
+
+use async_trait::async_trait;
+
+use iceberg::table::Table;
+use iceberg::Result;
+use iceberg::{
+    Catalog, Error, ErrorKind, Namespace, NamespaceIdent, TableCommit, TableCreation, TableIdent,
+};
+
+use crate::namespace_state::NamespaceState;
+
+/// In-memory catalog implementation.
+#[derive(Debug)]
+pub struct InMemoryCatalog {
+    root_namespace_state: Mutex<NamespaceState>,
+    file_io: FileIO,
+    default_table_root_location: String,
```

Comment:
`default_table_root_location` is part of `file_io` config, so we don't need to set this at catalog level.

Reply:
Apologies, I'm not understanding, could you be more specific here for me please? 

Where should I be getting this `default_table_root_location` value from when I'm inside the `InMemoryCatalog.create_table` method?

Reply:
I think we don't need to put `default_table_root_location` in `create_table`. Just create with given namespaces like:

```diff
  let location = format!(
-    "{}/{}/{}",
-    self.default_table_root_location,
+    "{}/{}",
     table_ident.namespace().join("/"),
     table_ident.name()
  );
```

Reply:
Ah okay, I see what you're saying now. 
CMIIW though isn't that going to be a bit problematic? 
You'll end up with files at `namespace/.*` of your file-system/s3-bucket/etc.? 
As opposed to with the current approach where you can specify `/tmp/namespace/.*` for example. 
I'm also thinking about unit-tests here, I don't want random files littering my hard drive. 
Let me know your thoughts.

Reply:
Based on my current understanding of the catalog API: If users want to configure the `location`, they should specify it in `TableCreate`. Otherwise, the location generation logic is the implementation detail.

I'm fine if all files been located under `<namespaces>/<uuids>`. If I want to control the `root`, I can setup it in `FileIO` directly.

Reply:
> For example, I can configure an fs FileIO with root = "/tmp".

CMIIW but I don't there is a way to do that today. 
I had to make a bunch of changes to enable that in the [latest commit](https://github.com/apache/iceberg-rust/pull/475/commits/807dd4cf649b5c367f25afc59f99341d6995c337). Please take another look.

Reply:
> CMIIW but I don't see a way to do that today.
> I had to make some changes to enable that (for the unit tests) in the latest commit.

Hi, it's better not to include unrelated changes in this PR. We can discuss about the default root in another issue.

Reply:
Fair, [reverted](https://github.com/apache/iceberg-rust/pull/475/commits/1d010721ee417cc38f56d7580871c2fcbf07b889). 

I've also removed `default_table_root_location` from `InMemoryCatalog` and refactored the unit tests to always explicitly provide a temporary table location.

Reply:
Thanks a lot! I think this PR can merge as is and I'm willing to do the changes.

Reply:
Sounds good. 

For posterity's sake, it would basically be just this [commit](https://github.com/apache/iceberg-rust/pull/475/commits/d778682361af6bd4fc2dcd3827736f1504fb5e54) to revert.

Reply:
How about starting a new thread instead. I think it's not a block of merge current PR.